### DISCUSSION TITLE
Downloading VSCode extensions through Nexus Repository

### DISCUSSION THREAD

**Discussion by Murt2003** [microsoft/vscode-discussions#1269] [Extension Development QnA] [answered]
Hello!

I am using a local Nexus Repository server (https://help.sonatype.com/en/sonatype-nexus-repository.html) to download R and Python packages and i'm looking into doing the same with Visual Studio Code for downloading extensions. This would help solve networking questions by going through the Nexus server, and also give the option to monitor what extensions are being used.

As I could not find a way myself, does anyone know if it is possible to use Nexus Repository in Visual Studio Code?

Thank you,
Mart

**Comment by theReynald** [comment, marked as answer]
Hi Murt2003,

According to GItHub Copilot, Visual Studio Code does not natively support configuring an alternative extension marketplace directly through its standard settings or configurations, which means you cannot directly configure it to use a Nexus Repository as a source for extensions in the same way you might for R and Python packages.

It goes on to say there are a few alternative approaches you might consider:

Manual Download and Installation: You can manually download VS Code extension .vsix files and distribute them within your organization. Users can then install these by using the Extensions: Install from VSIX... command in the Command Palette (Cmd+Shift+P on macOS or Ctrl+Shift+P on Windows/Linux).

Internal Extension Marketplace: Some organizations opt to create their own internal extension marketplace or repository. This would involve more complex setup and maintenance, including hosting your own server, managing extension updates, and possibly developing a custom extension or tool for easier integration with VS Code. This approach would allow you to have full control over which extensions are available and monitor their usage.

Proxy Configuration: For networking issues, you can configure VS Code (and many other tools) to use a proxy server. If your Nexus Repository is set up as a proxy, you can configure VS Code to route its internet traffic through this proxy. This doesn't directly affect where VS Code gets its extensions from but can help with networking issues and monitoring internet usage.

Scripting and Automation: You can use scripting and automation tools to manage the download and distribution of .vsix files from a central repository. This could involve automatically downloading updates for approved extensions and making them available on an internal server.

While these approaches can help manage and monitor VS Code extensions within an organization, they do come with additional overhead and complexity. It's important to evaluate the benefits against the effort required to implement and maintain these solutions.

Let me know if this helps.

/r

**Reply by Murt2003** [reply]
Thank you, theReynald, for this great and complete answer.

**Reply by okeiprivat55** [reply]
Hello i Thank you for all , Great Days.

**Comment by gabry801** [comment]
Hi theReynald, 
still not possible to use a "raw" or "proxy" repo on Nexus to get extensions?

**Comment by benji78** [comment]
I have looked into it and even tried to make it work as a "proxy", but VSCode's extension repository uses POST requests that are not possible through Sonatype Nexus because the response would change on every request while Nexus is supposed to be a _static file repository_.
Therefore the next best option is to use OpenVSX. You can install it in your organisation network and then set up open-vsx.org as it's upstream repository (I believe it is against Microsoft ToS to use their [extensions marketplace](https://marketplace.visualstudio.com/) for anything other than VSCode itself).
Once installed, you can configure a blacklist or whitelist. To set that OpenVSX repository as VSCode's  default (or VSCodium), there is an extension but I have not been able to make it work, so you need to change your `products.json` to something like this, changing `open-vsx.org` to your own domain:
```json
     "extensionsGallery": {
       "serviceUrl": "https://open-vsx.org/vscode/gallery",
       "itemUrl": "https://open-vsx.org/vscode/item",
       "extensionUrlTemplate": "https://open-vsx.org/vscode/gallery/{publisher}/{name}/latest",
       "controlUrl": "https://raw.githubusercontent.com/EclipseFdn/publish-extensions/refs/heads/master/extension-control/extensions.json"
     },
```
The `serviceUrl` is used to search extensions, and `itemUrl` redirects to extension detail pages when you click on an extension name in VS Code. VSCode may display a warning before opening a web browser to the repository's domain. In order to suppress these warnings, either edit the trusted domains with the corresponding button in the pop-up dialog, or add the following to your `product.json`:
```yaml
     "linkProtectionTrustedDomains": [
         "https://open-vsx.org"
     ]
```
Here is openVSX's documentation on this : https://github.com/eclipse/openvsx/wiki/Using-Open-VSX-in-VS-Code
